native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs

```diff
@@ -0,0 +1,178 @@
+
+namespace AWSGsrSerDe.serializer.json
+{
+    /// <summary>
+    /// Json serialization helper.
+    /// </summary>
+    public class JsonSerializer : IDataFormatSerializer
+    {
+        /// <inheritdoc />
+        public byte[] Serialize([NotNull]object data)
+        {
+            if (data is null)
+            {
+                throw new ArgumentNullException(nameof(data), "Cannot take null value when serialize");
+            }
+
+            var dataNode = GetDataNode(data);
+            var jsonSchema = GetSchema(data);
+
+            JsonValidator.ValidateDataWithSchema(jsonSchema, dataNode);
+
+            return WriteBytes(dataNode);
+        }
+
+        /// <inheritdoc />
+        public string GetSchemaDefinition([NotNull]object data)
+        {
+            if (data is null)
+            {
+                throw new ArgumentNullException(nameof(data), "Cannot get schema definition of null");
+            }
+
+            var schemaNode = GetSchemaNode(data);
+            return schemaNode.ToJsonString();
+        }
+
+        /// <inheritdoc />
+        public void Validate(string schemaDefinition, byte[] data)
+        {
+            // We assume the data bytes are encoded as UTF-8 Strings
+            var payload = System.Text.Encoding.UTF8.GetString(data);
+            var jsonDataWithSchema = JsonDataWithSchema.Build(schemaDefinition, payload);
+            Validate(jsonDataWithSchema);
+        }
+
+        /// <inheritdoc />
+        public void Validate(object data)
+        {
+            var schema = GetSchema(data);
+            var dataNode = GetDataNode(data);
+            JsonValidator.ValidateDataWithSchema(schema, dataNode);
+        }
+
+        /// <inheritdoc />
+        public void SetAdditionalSchemaInfo(object data, ref GlueSchemaRegistrySchema schema)
+        {
+            // No-op
+            // Currently we do not need to modify schema object for Json message
+        }
+
+        private static bool IsWrapper(object data)
+        {
+            return data is JsonDataWithSchema;
+        }
+
+        private static byte[] WriteBytes(JsonNode data)
+        {
+            var options = new JsonWriterOptions();
+            using var stream = new MemoryStream();
+            using var writer = new Utf8JsonWriter(stream, options);
+
+            data.WriteTo(writer, new JsonSerializerOptions());
+            writer.Flush();
+            var array = stream.ToArray();
+            return array;
+        }
+
+        private static JsonSchema GetSchema(object data)
+        {
+            try
+            {
+                if (IsWrapper(data))
+                {
+                    var dataObject = (JsonDataWithSchema)data;
+                    var getSchemaTask = JsonSchema.FromJsonAsync(dataObject.Schema);
+                    return getSchemaTask.Result;
+                }
+                var schema = data is string dataJsonString
+                    ? JsonSchema.FromSampleJson(dataJsonString)
```

Comment:
What's the significance of this?

Reply:
I am trying to support one more use case which is when the data object passed in is a plain json string. Do you think I should keep it?